### PR TITLE
Broken removeEventListener in unload sandbox (fix #400)

### LINKED ISSUE #400: Broken removeEventListener in unload sandbox
Check [src/client/sandbox/event/unload.js](https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/sandbox/event/unload.js#L89): function that you are passing in removeEventListener is a new another object, you need to store handler in order to remove it.

### PR DESCRIPTION
**Pull request by LavrovArtem**
\cc @inikulin @AndreyBelym 
@VasilyStrelyaev please check test name

### CODE REVIEW

File: test/client/fixtures/sandbox/event/unload-test.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+QUnit.testStart(function () {
+    iframeSandbox.on(iframeSandbox.RUN_TASK_SCRIPT, initIframeTestHandler);
+    iframeSandbox.off(iframeSandbox.RUN_TASK_SCRIPT, iframeSandbox.iframeReadyToInitHandler);
+});
+
+QUnit.testDone(function () {
+    iframeSandbox.off(iframeSandbox.RUN_TASK_SCRIPT, initIframeTestHandler);
+});
+
+asyncTest('BEFORE_UNLOAD_EVENT event must be called the last (GH-400)', function () {
```
Comment (VasilyStrelyaev) [inline test/client/fixtures/sandbox/event/unload-test.js]:
```
asyncTest('BEFORE_UNLOAD_EVENT must be called last (GH-400)', function () {
```

File: src/client/sandbox/event/unload.js [hunk 2]
```diff
@@ -56,6 +56,22 @@ export default class UnloadSandbox extends SandboxBase {
...
+
+        if (target.tagName && target.tagName.toLowerCase() === 'a') {
+            var href = nativeMethods.getAttribute.call(target, 'href');
+
+            this.isFakeIEBeforeUnloadEvent = /(^javascript:)|(^mailto:)|(^tel:)|(^#)/.test(href);
+        }
+    }
+
+    _reattachBeforeUnloadListener () {
+        // NOTE: reattach Listener, it'll be last in the queue.
```
Comment (VasilyStrelyaev) [inline src/client/sandbox/event/unload.js]:
```
        // NOTE: reattach the Listener, it'll be the last in the queue.
```

**Comment by VasilyStrelyaev** [comment]
\r-

File: src/client/sandbox/event/unload.js [hunk 3]
```diff
@@ -56,6 +56,22 @@ export default class UnloadSandbox extends SandboxBase {
         }
     }
 
+    _onDocumentClick (e) {
+        var target = e.target || e.srcElement;
+
+        if (target.tagName && target.tagName.toLowerCase() === 'a') {
+            var href = nativeMethods.getAttribute.call(target, 'href');
+
+            this.isFakeIEBeforeUnloadEvent = /(^javascript:)|(^mailto:)|(^tel:)|(^#)/.test(href);
```
Comment (inikulin) [inline src/client/sandbox/event/unload.js]: This could be any other pseudo-protocol. Shouldn't we check for `http/https` instead?

**Comment by inikulin** [comment]
\r-

**Comment by AndreyBelym** [comment]
lgtm

File: src/client/sandbox/event/unload.js [hunk 4]
```diff
@@ -56,6 +57,19 @@ export default class UnloadSandbox extends SandboxBase {
         }
     }
 
+    _onDocumentClick (e) {
+        var target = e.target || e.srcElement;
+
+        if (target.tagName && target.tagName.toLowerCase() === 'a')
+            this.isFakeIEBeforeUnloadEvent = !isSupportedProtocol(nativeMethods.getAttribute.call(target, 'href'));
```
Comment (inikulin) [inline src/client/sandbox/event/unload.js]: I've already told my concerns about `isSupportedProtocol`: https://github.com/DevExpress/testcafe-hammerhead/pull/348#discussion_r49736383 We should rewrite it or **stop** using it.

**Comment by inikulin** [comment]
\r-

**Comment by miherlosev** [comment]
see https://github.com/DevExpress/testcafe-hammerhead/issues/277#issuecomment-180276044

**Comment by LavrovArtem** [comment]
@miherlosev thanks

**Comment by LavrovArtem** [comment]
FPR
+
\cc @churkin @AlexanderMoskovkin

File: test/client/fixtures/sandbox/event/unload-test.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+                strictEqual(result, '12');
+
+                start();
+            });
+
+            window.addEventListener(window.unloadSandbox.beforeUnloadEventName, function () {
+                result += '1';
+            });
+
+            window.addEventListener(window.unloadSandbox.beforeUnloadEventName, function () {
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/sandbox/event/unload-test.js]: do we really need to add two handlers via `window.addEventListener`? It seems nothing is changed if we leave just one.
And should we check the handler added by `window.onbeforeunload|window.onpagehide` instead the second `addEventListener`?

**Comment by LavrovArtem** [comment]
FPR

**Comment by LavrovArtem** [comment]
ping @inikulin @churkin @AlexanderMoskovkin

**Comment by churkin** [comment]
\r-

**Comment by LavrovArtem** [comment]
FPR

**Comment by churkin** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by LavrovArtem** [comment]
ping @inikulin

**Comment by inikulin** [comment]
lgtm

**Pull request merged**